k2/csrc/fsa_utils.cu

```diff
@@ -2096,4 +2097,395 @@ void FixNumStates(FsaVec *fsas) {
+    cdf value for the next arc (taking it to be 1.0 if this is the last arc
+    leaving this state) will be exactly equal to the cdf value for this arc.
+
+    This makes easier the job of deciding, for some  0 <= p <= 1.0, which
+    arc leaving a specific state it "belongs to", without any danger of
+    picking an arc that goes to a state that cannot reach the final state.
+  */
+  FloatType *arc_cdf_data = arc_cdf.Data();
+  // `arc_inv_tots` will contain the inverses of the totals of the arc_post values
+  // leaving each state, if those totals were nonzero and finite; and otherwise, 1.0.
+  Array1<FloatType> arc_inv_tots(c, num_states);
+  FloatType *arc_inv_tots_data = arc_inv_tots.Data();
+  K2_EVAL(c, num_states, lambda_set_inv_tots, (int32_t i) {
+      int32_t begin_arc = fsas_row_splits2[i],
+          end_arc = fsas_row_splits2[i + 1];
+      FloatType inv_tot = 1.0;
+      if (end_arc > begin_arc) {
+        FloatType this_tot = arc_cdf_data[end_arc - 1] +
+            arc_pdf_data[end_arc - 1],
+            inv_tot;
+        if (this_tot > 0)
+          inv_tot = 1.0 / this_tot;
+        // we'll leave inv_tot at 1.0 for states that had zero or NaN
+        // `this_tot`, which could happen if those states were not accessible or
+        // not coaccessible.
+      }
+      arc_inv_tots_data[i] = inv_tot;
+    });
+  K2_EVAL(c, num_arcs, lambda_modify_cdf, (int32_t arc_idx012) {
+      int32_t state_idx01 = fsa_row_ids2[arc_idx012];
+      FloatType cdf_value = arc_cdf_data[arc_idx012],
+          cdf_value_modified = cdf_value * arc_inv_tots_data[state_idx01];
+      arc_cdf_data[arc_idx012] = cdf_value_modified;
+    });
+  return arc_cdf;
+}
+
+
+
+namespace random_paths_internal {
+
+
+// shared state (the changing part of it) for the algorithm that gets the paths.
+template <typename FloatType>
+struct PathState {
+  int32_t begin_arc_idx01x;  // first arc_idx012 leavng this state..
+  int32_t num_arcs;          // num arcs leaving this state.
+
+  // `p` is a number in the interval [0, 1], which you can think of as being
+  // random (although it's initialized deterministically at fixed intervals).
+  // As we advance `cur_state_idx01` from the start state to later states in the
+  // path, we zoom in more and more; and `p` is always the position within the
+  // probability interval spanned by arcs leaving `cur_state_idx01`.
+  FloatType p;
+};
+
+}  // namespace random_path_internal
+
+template <typename FloatType>
+Ragged<int32_t> RandomPaths(Ragged<FsaVec> &fsas,
+                            Array1<FloatType> &arc_cdf,
+                            const Array1<int32_t> &num_paths,
+                            Ragged<int32_t> &state_batches) {
+  using namespace random_path_internal;
+  K2_CHECK_EQ(fsas.NumAxes(), 3);
+  K2_CHECK_EQ(fsas.NumElements(), arc_cdf.Dim());
+  K2_CHECK_EQ(fsas.Dim0(), num_paths.Dim());
+  K2_CHECK(IsCompatible(fsas, arc_cdf));
```

Comment:
Just noted as we have defined `GetContext` long long ago, so it would be much simpler to use it for context checking stuff, i.e. `ContextPtr c = GetContext(fsas, arc_cdf, num_paths, state_batches)`